cli_tools/google-osconfig-agent/patch/patch_run.go

```diff
@@ -77,89 +109,251 @@ func (j *patchJob) UnmarshalJSON(b []byte) error {
-	logger.Debugf("patchrunner running %s", pr.Job.PatchJobName)
-	reboot := pr.run()
-	if pr.Job.PatchConfig.RebootConfig == osconfigpb.PatchConfig_NEVER {
+/**
+ * Runs a patch from start to finish. Sometimes this happens in a single invocation. Other times
+ * we need to handle the following edge cases:
+ * - The watcher has initiated this multiple times for the same patch job.
+ * - We have a saved state and are continuing after a reboot.
+ * - An error occurred and we do another attempt starting where we last failed.
+ * - The process was unexpectedly restarted and we are continuing from where we left off.
+ */
+func (r *patchRun) runPatch(ctx context.Context) {
+	savedPatchState, err := loadState(state)
+	if err != nil {
+		logger.Errorf("loadState error: %v", err)
 		return
 	}
-	if (pr.Job.PatchConfig.RebootConfig == osconfigpb.PatchConfig_ALWAYS) ||
-		(((pr.Job.PatchConfig.RebootConfig == osconfigpb.PatchConfig_DEFAULT) ||
-			(pr.Job.PatchConfig.RebootConfig == osconfigpb.PatchConfig_REBOOT_CONFIG_UNSPECIFIED)) &&
-			reboot) {
-		logger.Debugf("reboot requested %s", pr.Job.PatchJobName)
-		if err := rebootSystem(); err != nil {
-			logger.Errorf("error running reboot: %s", err)
+
+	if savedPatchState != nil && savedPatchState.Job.PatchJobName == r.Job.PatchJobName {
+		// continue from previous patch step
+		r.PatchStep = savedPatchState.PatchStep
+	} else {
+		// We have no saved state for this patch.
+		r.PatchStep = started
+	}
+
+	if r.PatchStep <= started {
+		logger.Debugf("Starting patchJob %s", r.Job)
+		r.StartedAt = time.Now()
+
+		if r.reportState(ctx, api.Instance_STARTED) {
+			return
+		}
+	}
+
+	if r.PatchStep <= prePatchReboot {
+		// check if we need to reboot
+		if r.rebootIfNeeded(ctx, false) {
+			return
+		}
+	}
+
+	if r.PatchStep <= applyPatch {
+		if r.reportState(ctx, api.Instance_APPLYING_PATCHES) {
+			return
+		}
+
+		if r.Job.DryRun {
+			logger.Infof("Dry run - No updates applied for patch job '%s'", r.Job.PatchJobName)
 		} else {
-			// Reboot can take a bit, shutdown the agent so other activities don't start.
-			os.Exit(0)
+			err = runUpdates(r.Job.PatchConfig)
+			if err != nil {
+				r.finishAndReportError(ctx, "Failed to apply patches: %v", err)
+				return
+			}
+		}
+	}
+
+	if r.PatchStep <= postPatchReboot {
+		// check if we need to reboot
+		if r.rebootIfNeeded(ctx, true) {
+			return
+		}
+	}
+
+	if r.PatchStep <= completeSuccess {
+		r.PatchStep = completeSuccess
+		r.reportSucceeded(ctx)
+	}
+	logger.Debugf("Completed patchJob %s", r.Job)
+}
+
+func patchRunner(ctx context.Context, pr *patchRun) {
+	logger.Debugf("Running patch job %s", pr.Job.PatchJobName)
+	pr.runPatch(ctx)
+	logger.Debugf("Finished patch job %s", pr.Job.PatchJobName)
+}
+
+func ackPatch(ctx context.Context, patchJobName string) {
+	currentPatchJob, err := loadState(state)
+	if err != nil {
+		logger.Errorf("Unable to load state to ack notification. Error: %v", err)
+		return
+	}
+
+	// Notify the server if we haven't yet
```

Comment:
Change this logic to not enqueue if the patch is already acked

Also add a TODO to store the already acked patch runs in memory so we don't have to read the state file.